File: app/src/main/java/net/xzos/UpgradeAll/ui/activity/HubListActivity.java
package net.xzos.UpgradeAll.ui.activity;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import net.xzos.UpgradeAll.R;
import net.xzos.UpgradeAll.database.HubDatabase;
import net.xzos.UpgradeAll.json.cache.ItemCardViewExtraData;
import net.xzos.UpgradeAll.server.hub.HubManager;
import net.xzos.UpgradeAll.ui.viewmodels.adapters.LocalHubItemAdapter;
import net.xzos.UpgradeAll.ui.viewmodels.view.ItemCardView;

import java.util.ArrayList;
import java.util.List;

import io.github.kobakei.materialfabspeeddial.FabSpeedDial;

public class HubListActivity extends AppCompatActivity {
    private List<ItemCardView> itemCardViewList = new ArrayList<>();
    private LocalHubItemAdapter adapter;

    private RecyclerView recyclerView;
    private SwipeRefreshLayout swipeRefresh;

    @Override
    protected void onResume() {
        super.onResume();
        refreshHubList();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_hub);
        // toolbar 点击事件
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
        }
        // tab添加事件
        FabSpeedDial fab = findViewById(R.id.addFab);
        fab.addOnMenuItemClickListener((floatingActionButton, textView, integer) -> {
            if (floatingActionButton == fab.getMiniFab(0)) {
                startActivity(new Intent(this, HubLocalActivity.class));
            } else if (floatingActionButton == fab.getMiniFab(1)) {
                startActivity(new Intent(HubListActivity.this, CloudHubListActivity.class));
            }
            return null;
        });

        swipeRefresh = findViewById(R.id.swipeRefresh);
        swipeRefresh.setColorSchemeResources(R.color.colorPrimary);
        swipeRefresh.setOnRefreshListener(this::refreshCardView);

        recyclerView = findViewById(R.id.update_item_recycler_view);

        setRecyclerView();
    }

    private void refreshCardView() {
        new Thread(() -> runOnUiThread(() -> {
            swipeRefresh.setRefreshing(true);
            refreshHubList();
            swipeRefresh.setRefreshing(false);
        })).start();
    }

    private void refreshHubList() {
        List<HubDatabase> hubDatabase = HubManager.getDatabases();
        itemCardViewList.clear();
        for (HubDatabase hubItem : hubDatabase) {
            String name = hubItem.getName();
            String uuid = hubItem.getUuid();
            itemCardViewList.add(new ItemCardView.Builder(name, uuid, "")
                    .extraData(new ItemCardViewExtraData.Builder().uuid(uuid).build())
                    .build());
        }
        TextView guidelinesTextView = findViewById(R.id.guidelinesTextView);
        if (itemCardViewList.size() != 0) {
            itemCardViewList.add(new ItemCardView.Builder(null, null, null)
                    .extraData(new ItemCardViewExtraData.Builder().isEmpty(true).build())
                    .build());
            setRecyclerView();
            adapter.notifyDataSetChanged();
            guidelinesTextView.setVisibility(View.GONE);
        } else {
            guidelinesTextView.setVisibility(View.VISIBLE);
        }
    }

    private void setRecyclerView() {
        GridLayoutManager layoutManager = new GridLayoutManager(this, 1);
        recyclerView.setLayoutManager(layoutManager);
        adapter = new LocalHubItemAdapter(itemCardViewList);
        recyclerView.setAdapter(adapter);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_actionbar_hub, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        int id = item.getItemId();

        switch (id) {
            case R.id.app_help:
                Intent intent = new Intent(Intent.ACTION_VIEW);
                intent.setData(Uri.parse("https://xzos.net/upgradeall-developer-documentation/"));
                intent = Intent.createChooser(intent, "请选择浏览器以查看帮助文档");
                startActivity(intent);
                return true;
            case android.R.id.home:
                onBackPressed();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

}

File: app/src/main/kotlin/net/xzos/upgradeAll/ui/activity/HubListActivity.kt
package net.xzos.upgradeAll.ui.activity

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import io.github.kobakei.materialfabspeeddial.FabSpeedDial
import net.xzos.upgradeAll.R
import net.xzos.upgradeAll.json.cache.ItemCardViewExtraData
import net.xzos.upgradeAll.server.hub.HubManager
import net.xzos.upgradeAll.ui.viewmodels.adapters.LocalHubItemAdapter
import net.xzos.upgradeAll.ui.viewmodels.view.ItemCardView
import java.util.*

class HubListActivity : AppCompatActivity() {
    private val itemCardViewList = ArrayList<ItemCardView>()
    private lateinit var adapter: LocalHubItemAdapter

    private lateinit var recyclerView: RecyclerView
    private lateinit var swipeRefresh: SwipeRefreshLayout

    override fun onResume() {
        super.onResume()
        refreshHubList()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_hub)
        // toolbar 点击事件
        val actionBar = supportActionBar
        actionBar?.setDisplayHomeAsUpEnabled(true)
        // tab添加事件
        val fab = findViewById<FabSpeedDial>(R.id.addFab)
        fab.addOnMenuItemClickListener { floatingActionButton, _, _ ->
            if (floatingActionButton === fab.getMiniFab(0)) {
                startActivity(Intent(this, HubLocalActivity::class.java))
            } else if (floatingActionButton === fab.getMiniFab(1)) {
                startActivity(Intent(this@HubListActivity, CloudHubListActivity::class.java))
            }
        }

        swipeRefresh = findViewById(R.id.swipeRefresh)
        swipeRefresh.setColorSchemeResources(R.color.colorPrimary)
        swipeRefresh.setOnRefreshListener { this.refreshCardView() }

        recyclerView = findViewById(R.id.update_item_recycler_view)

        setRecyclerView()
    }

    private fun refreshCardView() {
        Thread {
            runOnUiThread {
                swipeRefresh.isRefreshing = true
                refreshHubList()
                swipeRefresh.isRefreshing = false
            }
        }.start()
    }

    private fun refreshHubList() {
        val hubDatabase = HubManager.databases
        itemCardViewList.clear()
        for (hubItem in hubDatabase) {
            val name = hubItem.name
            val uuid = hubItem.uuid
            itemCardViewList.add(ItemCardView(name, uuid, "", ItemCardViewExtraData(uuid = uuid)))
        }
        val guidelinesTextView = findViewById<TextView>(R.id.guidelinesTextView)
        if (itemCardViewList.size != 0) {
            itemCardViewList.add(ItemCardView(null, null, null, ItemCardViewExtraData(isEmpty = true)))
            setRecyclerView()
            adapter.notifyDataSetChanged()
            guidelinesTextView.visibility = View.GONE
        } else {
            guidelinesTextView.visibility = View.VISIBLE
        }
    }

    private fun setRecyclerView() {
        val layoutManager = GridLayoutManager(this, 1)
        recyclerView.layoutManager = layoutManager
        adapter = LocalHubItemAdapter(itemCardViewList)
        recyclerView.adapter = adapter
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.menu_actionbar_hub, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.app_help -> {
                var intent = Intent(Intent.ACTION_VIEW)
                intent.data = Uri.parse("https://xzos.net/upgradeall-developer-documentation/")
                intent = Intent.createChooser(intent, "请选择浏览器以查看帮助文档")
                startActivity(intent)
                true
            }
            android.R.id.home -> {
                onBackPressed()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

}
